fix: remove oci:// prefix from registry host when login

alita1991 opened the pull request:
Closes #8

nflaig commented:
Do you have a reference on why this is all of the sudden needed with helm version > 3.10?

This solution feels a bit hacky to me. Based on the helm version the registry needs to be defined in a different format with and without `oci://` prtefix. Maybe it would be better to handle this internally so there is no difference on how the registry URI needs to be set. See https://github.com/nflaig/semantic-release-helm/issues/8#issuecomment-1489241000

Maybe @lsoica can take a look as well and give feedback

alita1991 commented:
Hi @nflaig 

You can read about this at https://helm.sh/docs/topics/registries/

This issue is also similar with mine https://github.com/helm/helm/issues/10814

nflaig commented:
thanks for the links, looks like helm just doesn't care about following semver

I really think this should be implemented differently by updating the [push command](https://helm.sh/docs/topics/registries/#the-push-subcommand) else once users upgrade to helm versions >= 3.8.0 their setup will break because they will have their `registry` URI set without `oci://` prefix.

![image](https://user-images.githubusercontent.com/38436224/228673184-aa6b4bd6-d5f7-4467-981f-d1d2a22aba19.png)

alita1991 commented:
@nflaig the change is only required when we log in to the registry, after this, we just have to use the registry defined in the config.
Do you think the change I made will impact also other commands like push?

nflaig commented:
I main concern is just that once others users upgrade to helm version 3.8.0 their setup will break because they did not explicity specify `oci://` in their registry URI

alita1991 commented:
Make sense, give me some time to update the code, will let you know

nflaig commented:
Do you by any chance know if helm push breaks in version 3.7.0 if you specify the `oci://` prefix? If that is not the case we can just always add it on push, and just for the sake of it always remove it on login

alita1991 commented:
Hi, I found this post https://blog.devgenius.io/helm-3-7-changes-to-oci-registry-16f2887e0704
I believe this answers your question, it will not break in 3.7.0, we are safe.

nflaig commented on lib/publish.js:
```diff
@@ -76,6 +76,7 @@ async function publishChartToRegistry(configPath, registry, name, version) {
                 );
 
                 const chartArchive = stdout.split(":")[1].trim();
+                registry = registry.includes("oci://") ? registry : "oci://".concat(registry)
```
`registry.includes("oci://")` will never be `true`, to be honest I liked stripping it off on the login operation because some people might specify it based on the updated helm docs

prefer not to use concat for appending string

```suggestion
                registry = registry.includes("oci://") ? registry : "oci://" + registry
```

or 

```suggestion
                registry = registry.includes("oci://") ? registry : `oci://${registry}`
```

alita1991 replied:
done

alita1991 commented:
If I modify the pluginConfig.registry in the verifyConditions.js default function, will remain modified in the other scripts as well? like publish.js?

alita1991 commented:
@nflaig please check, I made more changes

nflaig commented on lib/publish.js:
```diff
@@ -76,6 +76,7 @@ async function publishChartToRegistry(configPath, registry, name, version) {
                 );
 
                 const chartArchive = stdout.split(":")[1].trim();
+                registry = `oci://${registry}`
```
now we break setups where the `registry` starts with `oci://`

alita1991 replied:
reverted to the original commit

nflaig commented:
> If I modify the pluginConfig.registry in the verifyConditions.js default function, will remain modified in the other scripts as well? like publish.js?

No string are not mutated like this in javascript, else the `.split` would be a issue already

alita1991 commented:
@nflaig I believe it should be fine now

- if oci:// is provided, removed before login and not added because is already in place
- if oci:// is not provided, nothing to be removed before login and added before we do the push

nflaig commented on lib/publish.js:
```diff
@@ -76,6 +76,7 @@ async function publishChartToRegistry(configPath, registry, name, version) {
                 );
 
                 const chartArchive = stdout.split(":")[1].trim();
+                registry = registry.startsWith("oci://") ? registry : `oci://${registry}`
```
```suggestion
                registry = registry.startsWith("oci://") ? registry : `oci://${registry}`;
```

nflaig reviewed (approved):
lgtm, thanks @alita1991 for fixing this!

nflaig commented:
Published [semantic-release-helm3@2.4.2](https://www.npmjs.com/package/semantic-release-helm3/v/2.4.2)

Ruchira-R commented:
Helm version : version.BuildInfo{Version:"v3.9.3", GitCommit:"414ff28d4029ae8c8b05d62aa06c7fe3dee2bc58", GitTreeState:"clean", GoVersion:"go1.17.13"}



`Error: scheme prefix missing from remote (e.g. "oci://")`

This is still an issue. I am pushing my Helm Charts to Amazon ECR using oci://

Example : 
`helm push test-helm-chart.tgz oci://<AWS Account>.dkr.ecr.<region>.amazonaws.com/`
